### PR TITLE
 Adding Support for oneOf, anyOf and allOf

### LINKED ISSUE #1012: HCAGenerator cannot generation sample_bundle.json
HCAGenerator is unable to generate "https://raw.githubusercontent.com/HumanCellAtlas/metadata-schema/4.6.0/json_schema/sample_bundle.json" because it uses `oneOf`. This logic needs to be added into JSONgenerator

### PR DESCRIPTION
**Pull request by Bento007**
### Test plan
added test cases for anyOf, allOf, oneOf, and patternProperties.

### Deployment instructions & migrations -->
None.

### Release notes
- added oneOf, allOf, and anyOf support
- Fixed issues with json regex and ‘.’
- Simplify multipleOf for integer and numbers
- Fixes #1012 

@hannes-ucsc  please review

### CODE REVIEW

File: tests/json_gen/generator.py [hunk 1]
```diff
@@ -29,6 +31,57 @@ def _check_value_types(self, value_types) -> tuple:
         return value_types
 
 
+def update(d, u):
```
Comment (hannes-ucsc) [inline tests/json_gen/generator.py]: Consider renaming to `_update` if this is a method that's only used internally.

Comment (hannes-ucsc) [inline tests/json_gen/generator.py]: Needs docstring. It's not readily apparent what this is supposed to do.

File: tests/json_gen/generator.py [hunk 2]
```diff
@@ -29,6 +31,57 @@ def _check_value_types(self, value_types) -> tuple:
...
+            if 'min' in k:
+                d[k] = max(v, d.get(k, v))
+            elif 'max' in k:
+                d[k] = min(v, d.get(k, v))
+            else:
+                d[k] = v
+    return d
+
+
+def difference(schema, chosen):
```
Comment (hannes-ucsc) [inline tests/json_gen/generator.py]: ditto (docstring and rename)

File: tests/json_gen/generator.py [hunk 3]
```diff
@@ -29,6 +31,57 @@ def _check_value_types(self, value_types) -> tuple:
...
+            elif isinstance(schema_value, list):
+                schema[key] = [i for i in schema_value if i not in chosen_value]
+            else:
+                remove_keys.append(key)
+    for key in remove_keys:
+        schema.pop(key)
+    return schema
+
+
+def remove(d, u):
```
Comment (hannes-ucsc) [inline tests/json_gen/generator.py]: ditto (docstring and rename)

File: tests/json_gen/generator.py [hunk 4]
```diff
@@ -29,6 +31,57 @@ def _check_value_types(self, value_types) -> tuple:
...
+                    d['properties'].pop(req, None)
+                d[k] = [i for i in d[k] if i not in v]
+            elif isinstance(v, dict):
+                d[k] = remove(dv, v)
+            elif isinstance(v, list):
+                d[k] = [i for i in dv if i not in v]
+    return d
+
+
+def sanitize(regex: str) -> str:
```
Comment (hannes-ucsc) [inline tests/json_gen/generator.py]: I don't fully understand the purpose of this. You need to make a habit of explaining non-obvious things. This is only used once and seems to have a very specific purpose. Make that cohesion explicit by moving this function closer to the call site, potentially as a nested function.

A few more notes: `.` becomes a literal character inside brackets. What about dots that are already escaped? Last but not least take a look at re.escape(). All this under the preamble that I don't actually understand the issue this is supposed to address.

**Comment by Bento007** [comment]
Monday it is.

File: tests/json_gen/generator.py [hunk 5]
```diff
@@ -29,6 +31,74 @@ def _check_value_types(self, value_types) -> tuple:
...
+    """
+    Updates an existing JSON schema. If the item is a list it is appended with the new values. If the item is a dict it
+    is updated with new keys. All other values are replaced with updated values. A special case for keys with min and
+    max in the name exists, where only the min or max respectively, is retained between the new and updated value.
+
+    :param target: The schema to apply the updates on.
+    :param updates: The schema to be applied.
+    """
+    for k, v in updates.items():
+        if isinstance(v, dict):
```
Comment (Bento007) [inline tests/json_gen/generator.py]: check instance types match

File: tests/json_gen/generator.py [hunk 6]
```diff
@@ -29,6 +31,74 @@ def _check_value_types(self, value_types) -> tuple:
...
+            if 'min' in k:
+                target[k] = max(v, target.get(k, v))
+            elif 'max' in k:
+                target[k] = min(v, target.get(k, v))
+            else:
+                target[k] = v
+    return target
+
+
+def _difference(provided: dict, chosen: dict) -> dict:
```
[the remark this answers is not in the document]
Comment (Bento007) [reply tests/json_gen/generator.py]: try to conditionally add to the result.

File: tests/json_gen/generator.py [hunk 7]
```diff
@@ -96,11 +166,30 @@ def _gen_json(self, schema: dict):
...
-                impostor = getattr(self, f"_{json_type}")(schema)
+                temp_schema = deepcopy(schema)
+                all_of = temp_schema.get('allOf')
+                if all_of is not None:
+                    for subschema in all_of:
+                        _update(temp_schema, subschema)
+                any_of = temp_schema.get('anyOf')
+                if any_of is not None:
+                    for subschema in random.sample(any_of, random.randint(1, len(any_of) - 1)):
+                        _update(temp_schema, subschema)
```
Comment (Bento007) [inline tests/json_gen/generator.py]: Mention that there can be contradictions in any off that produce an invalid schema.

File: tests/json_gen/generator.py [hunk 8]
```diff
@@ -96,11 +166,30 @@ def _gen_json(self, schema: dict):
...
                 impostor = self._ref(ref)
             else:
-                impostor = getattr(self, f"_{json_type}")(schema)
+                temp_schema = deepcopy(schema)
+                all_of = temp_schema.get('allOf')
+                if all_of is not None:
+                    for subschema in all_of:
+                        _update(temp_schema, subschema)
+                any_of = temp_schema.get('anyOf')
+                if any_of is not None:
```
Comment (Bento007) [inline tests/json_gen/generator.py]: just pick one to avoid contradictions.

File: tests/json_gen/generator.py [hunk 9]
```diff
@@ -96,11 +166,30 @@ def _gen_json(self, schema: dict):
...
+                    for subschema in random.sample(any_of, random.randint(1, len(any_of) - 1)):
+                        _update(temp_schema, subschema)
+                one_of = temp_schema.get('oneOf')
+                if one_of is not None:
+                    subschema_choice = random.choice(one_of)
+                    _update(temp_schema, subschema_choice)
+                    remove_subschema = {}  # type: Dict[str, Any]
+                    for subschema in one_of:
+                        if subschema is not subschema_choice:
+                            _update(remove_subschema, _difference(subschema, subschema_choice))
```
Comment (Bento007) [inline tests/json_gen/generator.py]: don't modify values in place in _update, _difference, and _ removes.

File: tests/json_gen/generator.py (hunk 5 above)
Comment (Bento007) [reply tests/json_gen/generator.py]: done

File: tests/json_gen/generator.py (hunk 8 above)
Comment (Bento007) [reply tests/json_gen/generator.py]: done

File: tests/json_gen/generator.py (hunk 7 above)
Comment (Bento007) [reply tests/json_gen/generator.py]: only chose one, also added a comment

**Comment by Bento007** [comment]
Changes have been merged into dcp-pyjsongen https://github.com/HumanCellAtlas/dcp-pyjsongen/commit/21f75d8a54d624b5885236c6b86f259be943be36

**Pull request closed without merging**